source/common/router/router.cc

```diff
@@ -1333,16 +1334,37 @@ Filter::UpstreamRequest::UpstreamRequest(Filter& parent, Http::ConnectionPool::I
         parent_.callbacks_->tracingConfig(), "router " + parent.cluster_->name() + " egress",
         parent.timeSource().systemTime());
     span_->setTag(Tracing::Tags::get().Component, Tracing::Tags::get().Proxy);
+    span_->setTag(Tracing::Tags::get().UpstreamCluster, parent.cluster_->name());
+    span_->setTag(Tracing::Tags::get().HttpProtocol,
+                  AccessLog::AccessLogFormatUtils::protocolToString(pool.protocol()));
+    if (parent.attempt_count_ != 1) {
+      // This is a retry request, add this metadata to span.
+      span_->setTag(Tracing::Tags::get().RetryCount, std::to_string(parent.attempt_count_ - 1));
+    }
   }
 
   stream_info_.healthCheck(parent_.callbacks_->streamInfo().healthCheck());
 }
 
 Filter::UpstreamRequest::~UpstreamRequest() {
   if (span_ != nullptr) {
-    // TODO(mattklein123): Add tags based on what happened to this request (retries, reset, etc.).
+    if (parent_.grpc_request_) {
+      // Add gRPC response code to span.
+      const auto response_code =
+          Grpc::Common::responseToGrpcStatus(stream_info_, *upstream_headers_, *upstream_trailers_);
```

Comment:
I don't think upstream_trailers_ is guaranteed to be non-null here. You need to handle this case. (Please add relevant tests.)